Closing a process's stdout when there's more data to be read results in internal asyncio error

When using anyio with asyncio (This works fine with trio), if a subprocess is started with piped stdout and the stdout is closed (either explicitly or via context manager) before all data is read, asyncio will crash with a horrible traceback.

MVCE:

```python
import sys
import anyio

CODE = """\
import sys
for _ in range(1000):
    sys.stdout.buffer.write(bytes(range(256)))
"""

async def amain():
    async with await anyio.open_process([sys.executable, "-c", CODE]) as p:
        print(repr(await p.stdout.receive()))
        await p.stdout.aclose()  # The error even happens if you comment out this line!

if __name__ == "__main__":
    anyio.run(amain)
```

<details>
<summary>Output (after skipping the <code>print</code>'s output)</summary>

```
handle: <Handle SubprocessStreamProtocol.pipe_data_received(1, b'\x00\x01\x0...c\xfd\xfe\xff')>
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/subprocess.py", line 72, in pipe_data_received
    reader.feed_data(data)
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/streams.py", line 456, in feed_data
    assert not self._eof, 'feed_data after feed_eof'
AssertionError: feed_data after feed_eof
Exception in callback SubprocessStreamProtocol.pipe_data_received(1, b'\x00\x01\x0...c\xfd\xfe\xff')
handle: <Handle SubprocessStreamProtocol.pipe_data_received(1, b'\x00\x01\x0...c\xfd\xfe\xff')>
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/subprocess.py", line 72, in pipe_data_received
    reader.feed_data(data)
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/streams.py", line 456, in feed_data
    assert not self._eof, 'feed_data after feed_eof'
AssertionError: feed_data after feed_eof
Exception in callback SubprocessStreamProtocol.pipe_data_received(1, b'\x00\x01\x0...c\xfd\xfe\xff')
handle: <Handle SubprocessStreamProtocol.pipe_data_received(1, b'\x00\x01\x0...c\xfd\xfe\xff')>
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/subprocess.py", line 72, in pipe_data_received
    reader.feed_data(data)
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/streams.py", line 456, in feed_data
    assert not self._eof, 'feed_data after feed_eof'
AssertionError: feed_data after feed_eof
Exception in callback SubprocessStreamProtocol.pipe_data_received(1, b'\x00\x01\x0...c\xfd\xfe\xff')
handle: <Handle SubprocessStreamProtocol.pipe_data_received(1, b'\x00\x01\x0...c\xfd\xfe\xff')>
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/subprocess.py", line 72, in pipe_data_received
    reader.feed_data(data)
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/streams.py", line 456, in feed_data
    assert not self._eof, 'feed_data after feed_eof'
AssertionError: feed_data after feed_eof
Exception in callback SubprocessStreamProtocol.pipe_data_received(1, b'\x00\x01\x0...c\xfd\xfe\xff')
handle: <Handle SubprocessStreamProtocol.pipe_data_received(1, b'\x00\x01\x0...c\xfd\xfe\xff')>
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/subprocess.py", line 72, in pipe_data_received
    reader.feed_data(data)
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/streams.py", line 456, in feed_data
    assert not self._eof, 'feed_data after feed_eof'
AssertionError: feed_data after feed_eof
```

</details>

This behavior has been observed in the following environments:

* macOS 11.6.8, Python 3.10.8, anyio 3.6.2
* Debian 11 (bullseye), Python 3.8.13, anyio 3.6.1

--- a/docs/versionhistory.rst
+++ b/docs/versionhistory.rst
@@ -13,6 +13,9 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
 - Fixed ``to_process.run_sync()`` failing to initialize if ``__main__.__file__`` pointed
   to a file in a nonexistent directory
   (`#696 <https://github.com/agronholm/anyio/issues/696>`_)
+- Fixed ``AssertionError: feed_data after feed_eof`` on asyncio when a subprocess is
+  closed early, before its output has been read
+  (`#490 <https://github.com/agronholm/anyio/issues/490>`_)
 
 **4.4.0**
 

--- a/src/anyio/_backends/_asyncio.py
+++ b/src/anyio/_backends/_asyncio.py
@@ -23,7 +23,7 @@ from collections.abc import AsyncIterator, Generator, Iterable
 from concurrent.futures import Future
 from contextlib import suppress
 from contextvars import Context, copy_context
-from dataclasses import dataclass
+from dataclasses import dataclass, field
 from functools import partial, wraps
 from inspect import (
     CORO_RUNNING,
@@ -916,8 +916,12 @@ class BlockingPortal(abc.BlockingPortal):
 @dataclass(eq=False)
 class StreamReaderWrapper(abc.ByteReceiveStream):
     _stream: asyncio.StreamReader
+    _closed: bool = field(init=False, default=False)
 
     async def receive(self, max_bytes: int = 65536) -> bytes:
+        if self._closed:
+            raise ClosedResourceError
+
         data = await self._stream.read(max_bytes)
         if data:
             return data
@@ -925,7 +929,7 @@ class StreamReaderWrapper(abc.ByteReceiveStream):
             raise EndOfStream
 
     async def aclose(self) -> None:
-        self._stream.feed_eof()
+        self._closed = True
         await AsyncIOBackend.checkpoint()
 
 

--- a/src/anyio/_backends/_asyncio.py
+++ b/src/anyio/_backends/_asyncio.py
@@ -24,7 +24,7 @@ from collections.abc import AsyncIterator, Iterable
 from concurrent.futures import Future
 from contextlib import suppress
 from contextvars import Context, copy_context
-from dataclasses import dataclass, field
+from dataclasses import dataclass
 from functools import partial, wraps
 from inspect import (
     CORO_RUNNING,
@@ -908,12 +908,8 @@ class BlockingPortal(abc.BlockingPortal):
 @dataclass(eq=False)
 class StreamReaderWrapper(abc.ByteReceiveStream):
     _stream: asyncio.StreamReader
-    _closed: bool = field(init=False, default=False)
 
     async def receive(self, max_bytes: int = 65536) -> bytes:
-        if self._closed:
-            raise ClosedResourceError
-
         data = await self._stream.read(max_bytes)
         if data:
             return data
@@ -921,7 +917,7 @@ class StreamReaderWrapper(abc.ByteReceiveStream):
             raise EndOfStream
 
     async def aclose(self) -> None:
-        self._closed = True
+        self._stream.set_exception(ClosedResourceError())
         await AsyncIOBackend.checkpoint()
 
 

--- a/tests/test_subprocesses.py
+++ b/tests/test_subprocesses.py
@@ -10,7 +10,13 @@ from textwrap import dedent
 import pytest
 from _pytest.fixtures import FixtureRequest
 
-from anyio import CancelScope, ClosedResourceError, open_process, run_process
+from anyio import (
+    CancelScope,
+    ClosedResourceError,
+    create_task_group,
+    open_process,
+    run_process,
+)
 from anyio.streams.buffered import BufferedByteReceiveStream
 
 pytestmark = pytest.mark.anyio
@@ -237,3 +243,18 @@ async def test_close_early() -> None:
 
     async with await open_process([sys.executable, "-c", code]):
         pass
+
+
+async def test_close_while_reading() -> None:
+    code = dedent("""\
+    import sys
+    for _ in range(100):
+        sys.stdout.buffer.write(bytes(range(256)))
+    """)
+
+    async with await open_process(
+        [sys.executable, "-c", code]
+    ) as process, create_task_group() as tg:
+        tg.start_soon(process.stdout.aclose)
+        with pytest.raises(ClosedResourceError):
+            await process.stdout.receive()

--- a/tests/test_subprocesses.py
+++ b/tests/test_subprocesses.py
@@ -311,9 +311,9 @@ async def test_close_early() -> None:
 
 async def test_close_while_reading() -> None:
     code = dedent("""\
-    import sys
-    for _ in range(100):
-        sys.stdout.buffer.write(bytes(range(256)))
+    import time
+
+    time.sleep(3)
     """)
 
     async with await open_process(
@@ -323,3 +323,5 @@ async def test_close_while_reading() -> None:
         tg.start_soon(process.stdout.aclose)
         with pytest.raises(ClosedResourceError):
             await process.stdout.receive()
+
+        process.terminate()

--- a/tests/test_subprocesses.py
+++ b/tests/test_subprocesses.py
@@ -225,3 +225,15 @@ async def test_process_aexit_cancellation_closes_standard_streams(
 
     with pytest.raises(ClosedResourceError):
         await process.stderr.receive(1)
+
+
+async def test_close_early() -> None:
+    """Regression test for #490."""
+    code = dedent("""\
+    import sys
+    for _ in range(100):
+        sys.stdout.buffer.write(bytes(range(256)))
+    """)
+
+    async with await open_process([sys.executable, "-c", code]):
+        pass

